internal/slicesx/heap.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slicesx
+
+import "cmp"
+
+// Heap is a port of Go's [container/heap] package to use generics instead of
+// interface calls. Instead, entries consist of an ordered key and an arbitrary
+// value.ß
+//
+// A zero heap is empty and ready to use.
+type Heap[K cmp.Ordered, V any] struct {
```

Comment:
It would be nice to provide a constructor that can pre-allocate the keys and vals slices. For n-way merging, we could allocate a slice of length `n`.
```go
func NewHeapWithCapacity[K cmp.Ordered, V any](capacity int) *Heap[K,V] {
	return &Heap[K,V]{
		keys: make([]K, 0, capacity),
		vals: make([]V, 0, capacity),
	}
}
```